apps/builder/app/builder/features/style-panel/shared/css-value-input/unit-select.tsx

```diff
@@ -12,71 +16,187 @@ import {
   nestedSelectButtonUnitless,
 } from "@webstudio-is/design-system";
 import { ChevronDownIcon, ChevronUpIcon } from "@webstudio-is/icons";
+import type { CssValueInputValue } from "./css-value-input";
+import { isValid } from "../parse-css-value";
+import { toPascalCase } from "../keyword-utils";
 
-type UnitOption = {
-  id: Unit;
-  label: string;
-};
+type UnitOption =
+  | {
+      id: Unit;
+      label: string;
+      type: "unit";
+    }
+  | { id: string; label: string; type: "keyword" };
 
-const visibleLengthUnits = ["px", "em", "rem", "ch", "vw", "vh"] as const;
+// To make sorting stable
+const preferedSorting = [
+  "number",
+  "px",
+  ...units.percentage,
+  "em",
+  "rem",
+  "dvw",
+  "dvh",
+  ...units.length,
+  ...units.angle,
+  ...units.decibel,
+  ...units.flex,
+  ...units.frequency,
+  ...units.resolution,
+  ...units.semitones,
+  ...units.time,
+];
+
+const visibleLengthUnits = ["px", "em", "rem", "dvw", "dvh"] as const;
 
 type UseUnitSelectType = {
   property: string;
-  value?: Unit;
-  showUnitless: boolean;
-  onChange: (value: Unit) => void;
+  value: CssValueInputValue;
+  onChange: (
+    value: { type: "unit"; value: Unit } | { type: "keyword"; value: string }
+  ) => void;
   onCloseAutoFocus: (event: Event) => void;
 };
 
 export const useUnitSelect = ({
   property,
   value,
-  // edge-case, most css properties accept unitless value 0
-  showUnitless,
   onChange,
   onCloseAutoFocus,
 }: UseUnitSelectType): [boolean, JSX.Element | null] => {
   const [isOpen, setIsOpen] = useState(false);
 
+  const unit =
+    value.type === "unit" || value.type === "intermediate"
+      ? value.unit
+      : undefined;
+
   const options = useMemo(() => {
     const options: UnitOption[] = [];
     const { unitGroups } = properties[property as keyof typeof properties];
+
     for (const unitGroup of unitGroups) {
       if (unitGroup === "number") {
-        options.push({ id: "number", label: nestedSelectButtonUnitless });
+        options.push({
+          id: "number",
+          type: "unit",
+          label: nestedSelectButtonUnitless,
+        });
         continue;
       }
+
       const visibleUnits =
         unitGroup === "length" ? visibleLengthUnits : units[unitGroup];
       for (const unit of visibleUnits) {
-        options.push({ id: unit, label: unit.toLocaleUpperCase() });
+        options.push({
+          id: unit,
+          type: "unit",
+          label: unit.toLocaleUpperCase(),
+        });
       }
     }
 
-    if (showUnitless && !options.some((o) => o.id === "number")) {
-      options.push({ id: "number", label: nestedSelectButtonUnitless });
+    // Special case for 0, which is often used as a unitless value
+    const showUnitless =
```

Comment:
seems like majority of the logic in this hook could be a normal function which we could then test quite easily, wdyt?

Reply:
a little worried about all the use cases this hook handles inside of all those if statements, not sure what's tested here

Reply:
could be buildOptions()